src/snmalloc.h

```diff
@@ -7,10 +7,12 @@
 #include "mem/globalconfig.h"
 
 // The default configuration for snmalloc
+#ifndef SNMALLOC_PROVIDE_OWN_CONFIG
```

Comment:
This is what I'm doing.  Now that everything is templated nicely, there's no need to define things in the middle of the snmalloc includes: you can include everything then specify your config and then the type for it.  This is only necessary if you want to replace the *default* config, which is necessary only for using the override things.

Reply:
I think you need provide some information here.  My original intent was for any custom configuration to be
```
#include <snmalloc_core.h>

// Custom config that defines what Alloc means
...
namespace snmalloc
{
  using Alloc = ...
} // namespace snmalloc

#incude <snmalloc_front.h>
```

But with this I am not sure how to use this header.  Do I need to include something before 
```
#include <snmalloc.h>
```
If I include this header with `SNMALLOC_PROVIDE_OWN_CONFIG` set, then `snmalloc_front.h` won't compile as it doesn't have a definition of `snmalloc::Alloc`?

Reply:
In my `sandbox.h` I [include `snmalloc.h`](https://github.com/microsoft/verona/blob/bf705fa0dc443f2c0d8331c3ea5d17bf34d6dbed/experiments/process_sandbox/include/process_sandbox/sandbox.h#L22) and later define a config that provides the [sandbox definitions for the outside allocator](https://github.com/microsoft/verona/blob/bf705fa0dc443f2c0d8331c3ea5d17bf34d6dbed/experiments/process_sandbox/include/process_sandbox/sandbox.h#L323).  This is then [used directly as an explicit template instantiation](https://github.com/microsoft/verona/blob/bf705fa0dc443f2c0d8331c3ea5d17bf34d6dbed/experiments/process_sandbox/include/process_sandbox/sandbox.h#L387).  

When this file is included by thing in the host side, nothing I've written uses the definition of `Alloc`, but in the child I do and I depend on it being a custom thing and I also need to include `sandbox.h`, which includes `snmalloc.h`.

Without this change, if I include a header that includes `snmalloc.h` *after* I include the snmalloc headers that I need to define my custom malloc then I get a build failure because the `snmalloc.h` inclusion redefines `Alloc` with a different type.

With this change, it's a per-compilation-unit decision whether you want to override the default allocator definition and, if you do, then you must do so before you include `snmalloc.h` (which you may do indirectly).

Reply:
@davidchisnall if you included `snmalloc_core.h` does your example still work, and if not, what is it missing?

My view intent was roughly
* snmalloc_core.h  - things you need to build a custom snmalloc
* snmalloc_front.h - given a definition of snmalloc::Alloc give me the common things like thread local allocators, tests across the allocator pool (glocal_alloc.h), etc.
* snmalloc.h - default allocator setup for this platform.

Happy to change, but want to understand.

Reply:
I see.  I'd expect most users of {foo} to include {foo}.h.  I presumably could just include `snmalloc_core.h` in my `sandbox.h` instead of `snmalloc.h`, but I'm still a bit nervous if I add a library dependency and it includes `snmalloc.h` instead of `snmalloc_core.h` and that breaks my ability to define a custom `snmalloc::Alloc`.

Reply:
I guess my thought was that most people including `snmalloc.h` would just want a good allocator for their platform directly linked into the application.  

My concern with your current `#ifdef` is that it still includes `snmalloc_front.h`, which includes `ScopedAllocator` and `ThreadAlloc` both of which assume `snmalloc::Alloc` exists.  So I am surprised that this doesn't fail to compile. 

I also think `globalconfig.h` should not be included if `SNMALLOC_PROVIDE_OWN_CONFIG` is set.  As you might not want the statics to be compiled into your binary, if you are building your own config.

Perhaps, the right way to go is for `snmalloc.h` to degenerate to `snmalloc_core.h` if `SNMALLOC_PROVIDE_OWN_CONFIG` is set:

```C++
#pragma once

// Core implementation of snmalloc independent of the configuration mode
#include "snmalloc_core.h"

#ifndef SNMALLOC_PROVIDE_OWN_CONFIG
// Default implementation of global state
#  include "mem/globalconfig.h"

// The default configuration for snmalloc
namespace snmalloc
{
  using Alloc = snmalloc::LocalAllocator<snmalloc::Globals>;
}

// User facing API surface, needs to know what `Alloc` is.
#  include "snmalloc_front.h"
#endif
```

Then in we could perhaps eliminate `snmalloc_core` as it is not required, as you can always use `snmalloc.h`.  Would this be acceptable?  I think it would fix your concern,